### ISSUE TITLE
Field names contains dot

### ISSUE THREAD

**Issue by AnthonySteele** [justeat/NLog.StructuredLogging.Json#29]
Mykola Shestopal says:
> can we rename exception fields to not contain dots in them (for example Data.SomeField become Data_SomeField or so) in Structured Logging since this cause errors in Elastic Search so logs are not indexed if they have "Data" field (since earlier log had Data.SomeField so Data becomes an object and now needs to have properties, but Data field treated as usual field and can't be parsed as an object)
[12:45 PM] Mykola Shestopal: can this be configurable by some option, like rename fields with dot's (and which separator to use instead of dot)

See also:
https://github.com/elastic/elasticsearch/issues/2354
http://blog.endpoint.com/2013/04/elasticsearch-object-mapping-eof-400.html

**Comment by AnthonySteele** [comment]
Do we do this anywhere in StructuredLogging? I'm looking at [This code](https://github.com/justeat/NLog.StructuredLogging.Json/blob/master/src/NLog.StructuredLogging.Json/Helpers/Mapper.cs#L23) and [this code](https://github.com/justeat/NLog.StructuredLogging.Json/blob/master/src/NLog.StructuredLogging.Json/FlattenedJsonLayout.cs#L69), and not seeing any dots in field names. Can we get an example from the kibana logs of fields with dots in them, and see if it is generated by this library, or fixed by using this library, etc?

**Comment by AnthonySteele** [comment]
StructuredLogging.Json does not log exception details in this format, so the fix may be for the offending code to just use StructuredLogging.Json.
However, if you add custom fields containing dots or other special characters,  StructuredLogging.Json does not sanitize the field names in any way, so it is possible to abuse it.
I would favour docs and user education over adding field name sanitization code to StructuredLogging.Json, but this is also possible.

**Comment by brainmurphy** [comment]
There's nothing wrong with the actual file on disk though, right? Not sure I'd like to see this project grow workarounds for problems you might have if you decide to push those logs into ELK.

**Comment by AnthonySteele** [comment]
Pushing logs to ELK is pretty common (1 out of 1 users of this library do it) but I agree that StructuredLogging.Json  actively mangling all input field names to fit the specification of ELK is not ideal - it's crossing concerns, coupling this library to ELK, is likely to require a careful reading of how kibana treats field names to do just right (remember how terrible this idea is for valid email addresses), and will also require docs and user education anyway.

**Comment by martincostello** [comment]
Shall we close this if it's `wontfix`?

**Comment by snakefoot** [comment]
@AnthonySteele I think the problem is isolated to this method, that generates bad data when meeting bad exceptions (Microsoft SQL exceptions):

`private static void TransferDataObjectToLogEventProperties(LogEventInfo log, object logProperties)`

Not sure if this will fix the issue, but maybe someone can test?

```
            foreach (var dataProperty in properties)
            {
                string key = dataProperty.Key;
                if (key.IndexOf('.') > 0)
                     key = key.Replace('.', '_');
                log.Properties.Add(key, dataProperty.Value);
            }
```

Not sure extra work is needed for handling of innerexception that contains special Data-Dictionary-keys

**Comment by snakefoot** [comment]
See also:

https://www.elastic.co/guide/en/elasticsearch/reference/2.4/dots-in-names.html

**Comment by AnthonySteele** [comment]
To repeat what I said earlier, this library is not directly concerned with ElasticSearch, so should still work in cases where it is not involved, and should not couple to all of ElasticSearch's conventions. 

However if this data comes directly out of the exception, it seems that the removal of dots from field names would have to be done in here or in the infrastructure that forwards messages to ElasticSearch - i.e. `nxlog` or `Logstash`.

**Comment by snakefoot** [comment]
> This library is not directly concerned with ElasticSearch

Understand the argument with garbage-in and garbage-out. But Exceptions from external libraries (Like Microsoft SqlExceptions) are not under control from the application owner.

Pretty sure ElasticSearch is not the only application, that get confused by dots in the json-properties.

Anyway just wanted to point out it would not be a major code-change to support this (could be controlled with a parameter).

**Comment by AnthonySteele** [comment]
If this is just for the data dictionary of exceptions, then the right place to hook it in is [somewhere near here](https://github.com/justeat/NLog.StructuredLogging.Json/blob/master/src/NLog.StructuredLogging.Json/Helpers/Mapper.cs#L44)

**Comment by lukemerrett** [comment]
`Logstash` plugin for removing dots for reference https://www.elastic.co/guide/en/logstash/current/plugins-filters-de_dot.html

**Comment by AnthonySteele** [comment]
Fixed by  #93 for the "data dictionary of exceptions" case and similar.
